sdk/core/core-rest-pipeline/src/fetchHttpClient.ts

```diff
@@ -0,0 +1,237 @@
+} from "./interfaces";
+import { RestError } from "./restError";
+import { createHttpHeaders } from "./httpHeaders";
+
+/**
+ * Checks if the body is a NodeReadable stream which is not supported in Browsers
+ */
+function isNodeReadableStream(body: any): body is NodeJS.ReadableStream {
+  return body && typeof body.pipe === "function";
+}
+
+/**
+ * Checks if the body is a ReadableStream supported by browsers
+ */
+function isReadableStream(body: unknown): body is ReadableStream {
+  return Boolean(
+    body &&
+      typeof (body as ReadableStream).getReader === "function" &&
+      typeof (body as ReadableStream).tee === "function"
+  );
+}
+
+/**
+ * A HttpClient implementation that uses window.fetch to send HTTP requests.
+ * @internal
+ */
+class FetchHttpClient implements HttpClient {
+  /**
+   * Makes a request over an underlying transport layer and returns the response.
+   * @param request - The request to be made.
+   */
+  public async sendRequest(request: PipelineRequest): Promise<PipelineResponse> {
+    const url = new URL(request.url);
+    const isInsecure = url.protocol !== "https:";
+
+    if (isInsecure && !request.allowInsecureConnection) {
+      throw new Error(`Cannot connect to ${request.url} while allowInsecureConnection is false.`);
+    }
+
+    if (request.proxySettings) {
+      throw new Error("HTTP proxy is not supported in browser environment");
+    }
+
+    try {
+      return await makeRequest(request);
+    } catch (e) {
+      throw getError(e, request);
+    }
+  }
+}
+
+/**
+ * Sends a request
+ */
+async function makeRequest(request: PipelineRequest): Promise<PipelineResponse> {
+  const abortController = handleAbortSignal(request);
+  const headers = buildFetchHeaders(request.headers);
+
+  const requestBody = buildRequestBody(request);
+
+  const response = await fetch(request.url, {
+    body: requestBody,
+    method: request.method,
+    headers: headers,
+    signal: abortController.signal,
+    credentials: request.withCredentials ? "include" : "same-origin",
+    redirect: "manual",
+    cache: "no-store",
+  });
+
+  return buildPipelineResponse(response, request);
+}
+
+/**
+ * Creates a pipeline response from a Fetch response;
+ */
+async function buildPipelineResponse(httpResponse: Response, request: PipelineRequest) {
+  const headers = buildPipelineHeaders(httpResponse);
+  const response: PipelineResponse = {
+    request,
+    headers,
+    status: httpResponse.status,
+  };
+
+  const bodyStream = isReadableStream(httpResponse.body)
+    ? buildBodyStream(httpResponse.body, request.onDownloadProgress)
+    : httpResponse.body;
+
+  const responseStream = new Response(bodyStream);
+
+  if (
+    // Value of POSITIVE_INFINITY in streamResponseStatusCodes is considered as any status code
+    request.streamResponseStatusCodes?.has(Number.POSITIVE_INFINITY) ||
+    request.streamResponseStatusCodes?.has(response.status)
+  ) {
+    response.blobBody = responseStream.blob();
```

Comment:
We probably need some option here for "return the stream itself and not the blob or text version" right? Maybe we could use NEGATIVE_INFINITY for that? :)

Reply:
Discussed offline. Added an opt-in switch to get streams in browser. By default we give users a blob which loads the whole thing in memory as it is for xhr now